src/chainstate/burn/db/burndb.rs

```diff
@@ -843,674 +648,811 @@ impl BurnDB {
+    pub fn get_canonical_burn_chain_tip_stubbed(conn: &Connection) -> Result<BlockSnapshot, db_error> {
+        let pox_db = PoxDB::stubbed();
+        SortitionDB::get_last_snapshot(conn, &PoxIdentifier::stubbed(), &pox_db)
+            .map(|opt| opt.expect("CORRUPTION: No canonical burnchain tip"))
+    }
+
+    /// Get the canonical Stacks chain tip -- this gets memoized on the canonical burn chain tip.
+    pub fn get_canonical_stacks_chain_tip_hash_stubbed(conn: &Connection) -> Result<(BurnchainHeaderHash, BlockHeaderHash), db_error> {
+        let sn = SortitionDB::get_canonical_burn_chain_tip_stubbed(conn)?;
+
+        let stacks_block_hash = sn.canonical_stacks_tip_hash;
+        let burn_block_hash = sn.canonical_stacks_tip_burn_hash;
+        Ok((burn_block_hash, stacks_block_hash))
+    }
+
+
+    /// Get an accepted stacks block header in a fork whose chain tip has not yet committed
+    /// to it.
+    // PoX TODO: once we integrate with the StacksChainController, this logic will go away
+    fn get_accepted_stacks_block_pointer(conn: &Connection, tip_burn_header_hash: &BurnchainHeaderHash, stacks_block_hash: &BlockHeaderHash) -> Result<Option<AcceptedStacksBlockHeader>, db_error> {
+        let args : &[&dyn ToSql] = &[tip_burn_header_hash, stacks_block_hash];
+        query_row_panic(conn, "SELECT * FROM canonical_accepted_stacks_blocks WHERE tip_burn_block_hash = ?1 AND stacks_block_hash = ?2", args,
+                        || format!("BUG: the same Stacks block {} shows up twice or more in the same burn chain fork (whose tip is {})", stacks_block_hash, tip_burn_header_hash))
+    }
+
+    /// Add an accepted Stacks block to the canonical accepted stacks header table, to indicate
+    /// that it will be committed to by the next burn block added to the canonical chain tip.  Used
+    /// to identify Stacks blocks that get accepted in the mean time, so we can ensure that the
+    /// canonical burn chain tip always points to the canonical stacks chain tip.
+    // PoX TODO: once we integrate with the StacksChainController, this logic will go away 
+    fn insert_accepted_stacks_block_pointer(tx: &Transaction, tip_burn_header_hash: &BurnchainHeaderHash, burn_header_hash: &BurnchainHeaderHash, stacks_block_hash: &BlockHeaderHash, stacks_block_height: u64) -> Result<(), db_error> {
+        let args: &[&dyn ToSql] = &[tip_burn_header_hash, burn_header_hash, stacks_block_hash, &u64_to_sql(stacks_block_height)?];
+        tx.execute("INSERT OR REPLACE INTO canonical_accepted_stacks_blocks (tip_burn_block_hash, burn_block_hash, stacks_block_hash, block_height) VALUES (?1, ?2, ?3, ?4)", args)
+            .map_err(db_error::SqliteError)?;
+        Ok(())
+    }
+
+    /// Get the maximum arrival index for any known snapshot.
+    fn get_max_arrival_index(conn: &Connection) -> Result<u64, db_error> {
+        match conn.query_row("SELECT arrival_index FROM snapshots ORDER BY arrival_index DESC LIMIT 1", NO_PARAMS,
```

Comment:
Recommend using `SELECT MAX(arrival_index)` here -- it's more in-line with SQL best practices.  Also, is there an index on `snapshots` for `arrival_index`?

Reply:
Yep -- updated in d1e6a2270